README.md
[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/telecomsante/pdf-viewer)
![](https://img.shields.io/badge/polymer-2.x-blue.svg)
![](https://img.shields.io/badge/licence-ISC-brightgreen.svg)


# &lt;pdf-viewer&gt;

`pdf-viewer` is a simple viewer, without any decoration, based on the [Mozilla's pdfjs library](https://github.com/mozilla/pdfjs-dist). Buttons have to be created by the parent page or component.

the component is written using es6 syntax, and polymer 2.x.

> __Nota :__ the pdfjs library uses XMLHttpRequest to load the pdf documents. So, you may encountered CORS problem, if you try to load a document from another origin. A solution could be to use a proxy to serve your documents, you could take advantage to use for example the project [CORS Anywhere](https://github.com/Rob--W/cors-anywhere).

Features
 - Next/Previous page
 - Zoom
 - Polymer 2.x
 - Dynamically change document without page reloading
 - Works in IE11, IOS, Android (chrome)

![](demo.gif)

Basic usage :

```html
<pdf-viewer   
  src="https://www.w3.org/TR/1998/REC-html40-19980424/html40.pdf"
  initial-zoom="fit-width",
  mode="double",
  page="5"></pdf-viewer>
```

# Using in polymer app

the file `bower_components/pdfjs-dist/build/pdf.worker.min.js` must be added to `extradependencies` section into the `polymer.json` file.

If for some reasons, bower components are installed in another folder, be aware to pass as attribute `PDFJS_workerSrc` the full path to the `pdf.worker.min.js` file.

The component is licensed under the [ISC License](LICENSE.md)


CHANGELOG.md
## 2.2.0

  - fix issue #8 : On IE11, the document isn't centered into the page
  - improves performance : redraws the page only when it's necessary

## 2.1.1

  - fix issue #7 : Page Number is resetting to 1 when loading a document
  - use importPath to load the pdf worker script `pdf.worker.min.js`

## 2.1.0

 - update to PDFjs 2.0.426
 - adds events :
   - pdf-viewer-outrange: when trying to display a page that doesn't exist
   - pdf-viewer-render: when the page has been rendered
 - fix : this._PDF.getPage is not a function

## 2.0.1

 - add a property to declare the PDFJS worker script
 - update doc 

## 2.0.0

 - update to polymer 2.x

## 1.0.1

 - fix #1 : Drag'n drop to move pdf in the viewer not working in es5 version
 - README : adds a note about document loading

## 1.0.0

public release


pdf-viewer.js
Polymer({

  is: 'pdf-viewer',

  properties: {
    /**
     * url of the pdf file to display
     * @type {String}
     */
    src: {
      type: String,
      observer: '_srcChanged',
    },

    /**
     * The page number to display
     *
     * The first page is 1
     * @type {Number}
     */
    page: {
      type: Number,
      value: 1,
      observer: '_pageChanged',
      reflectToAttribute: true,
      notify: true,
    },

    /**
     * total number of pages in the document
     * @type {Number}
     */
    pages: {
      type: Number,
      readOnly: true,
      reflectToAttribute: true,
      notify: true,
    },

    /**
     * the initial zoom when opening a document and
     * when the mode change
     *
     * the available values are :
     *  - `fit` : whole page,
     *  - `fit-width` : full width
     *
     * @type {String}
     */
    initialZoom: {
      type: String,
      value: "fit",
    },

    /**
     * The zoom value
     * @type {Number}
     */
    zoom: {
      type: Number,
      value: 1,
      observer: '_zoomChanged'
    },

    /**
     * The zoom ratio used
     * @type {Numder}
     */
    zoomRatio: {
      type: Number,
      value: 1.25,
    },

    /**
     * The view mode of the document
     *
     * the available values are :
     *  - single : single page
     *  - double : double page side by side
     *
     * @type {String}
     */
    mode: {
      type: String,
      value: "single",
      observer: '_modeChanged',
    },

    _views: {
      type: Number,
      value: 1,
    },
    _PDF: {
      type: Object,
      value: () => { return {}; },
    },
    _pageRendering: {
      type: Boolean,
      value: false,
    },
    _pageNumPending: {
      type: Number,
      value: null,
    },
    _firstRender: {
      type: Boolean,
      value: false,
    },
    _pos: {
      type: Object,
      value: () => { return {x:0, y:0} },
    }
  },

  attached() {
    this._ro = new ResizeObserver((entries, observer) => {
      console.log('resize detect');
      if(this._firstRender) this._drawPage();
    });
    this._ro.observe(this);
  },

  detached() {
    delete this._ro;
  },

  _srcChanged() {
    this.firstRender = false;
    this.page = 1;
    PDFJS.getDocument(this.src)
     .then(pdf => {
       this._setPages(pdf.numPages);
       this._PDF = pdf;
       switch(this.initialZoom) {
         case "fit-width":
          this.fitWidth();
          break;
        default:
          this.fit();
       }
     });
  },

  _pageChanged(newValue, oldValue) {
    if(!this.pages) {
      this.page = 1;
    }
    if(newValue > this.pages ) this.page = this.pages;
    if(newValue < 1) this.page = 1;
    if(this._PDF) this._drawPage();
  },

  _zoomChanged: function _zoomChanged(newValue, oldValue) {
    if (!this.zoom) {this.zoom = 1;}
    this.zoom = newValue;
  },

  _modeChanged(newValue, oldValue) {
    switch (this.mode) {
      case 'double':
        this._views = 2;
        this.$.viewer2.hidden = false;
        break;
      default:
        this._views = 1;
        this.$.viewer2.hidden = true;
    }
    if(this._firstRender) {
      this._recenter();
      switch(this.initialZoom) {
        case "fit-width":
          this.fitWidth();
          break;
        default:
          this.fit();
       }
    }
  },

  /**
   * move to the next page
   */
  next() {
    if(this.page < this.pages) this.page += this._views;
  },

  /**
   * move to the previous page
   */
  previous() {
    if(this.page > 1) this.page -= this._views;
  },

  /**
   * display the document full width
   */
  fitWidth() {
    this._PDF.getPage(this.page)
    .then(page => {
      let viewport = page.getViewport(1);
      let rect = this.$.container.getBoundingClientRect();
      this.zoom = (rect.width-20)/(this._views * viewport.width);
      if(!this._firstRender) {
        viewport = page.getViewport(this.zoom);
        this._pos.x = (rect.width - this._views * viewport.width) / 2;
        this._pos.y = 10;
        this.$.viewer.style.transform = `translate(${this._pos.x}px,${this._pos.y}px )`;
      }
      this._drawPage();
    })
  },

  /**
   * display the whole page
   */
  fit() {
    this._PDF.getPage(this.page)
    .then(page => {
      let viewport = page.getViewport(1);
      let rect = this.$.container.getBoundingClientRect();
      this.zoom = Math.min((rect.width-20)/this._views * viewport.width, (rect.height-20)/viewport.height);
      if(!this._firstRender) {
        viewport = page.getViewport(this.zoom);
        this._pos.x = (rect.width - this._views * viewport.width) / 2;
        this._pos.y = (rect.height - viewport.height) / 2;
        this.$.viewer.style.transform = `translate(${this._pos.x}px,${this._pos.y}px )`;
      }
      this._drawPage();
    })
  },

  /**
   * zoom in
   */
  zoomin() {
    this.zoom = this.zoom * this.zoomRatio;
    this._drawPage();
  },

  /**
   * zoom out
   */
  zoomout() {
    this.zoom = this.zoom / this.zoomRatio;
    this._drawPage();
  },

  _renderView(view, pg) {
    return new Promise( (resolve, reject) => {
      this._PDF.getPage(pg)
      .then( page => {
        const viewport = page.getViewport(this.zoom);
        view.width = viewport.width;
        view.height = viewport.height;

        const context = view.getContext('2d');
        const renderContext = {
          canvasContext: context,
          viewport: viewport
        };
        resolve( page.render(renderContext));
      })
      .catch( (err) => {
        this._PDF.getPage(1)
          .then( page => {
            const viewport = page.getViewport(this.zoom);
            view.width = viewport.width;
            view.height = viewport.height
          })
          .then( resolve );
      });
    })
  },

  _drawPage(pg) {
    if(!pg) pg = this.page;
    if(this._pageRendering) {
      this._pageNumPending = pg
      return;
    }
    this._pageRendering = true;
    this.$.spinner.active = true;
    const promises = [];
    if( this.mode === "single" ) {
      promises.push(this._renderView(this.$.viewer1, pg))
    } else {

      promises.push(this._renderView(this.$.viewer1, pg - pg % 2));
      promises.push(this._renderView(this.$.viewer2, pg - pg % 2 +1));
    }
    Promise.all(promises)
      .then( () => {
        this._firstRender = true;
        this._pageRendering = false;
        this.$.spinner.active = false;
        this._recenter();
        if (this._pageNumPending !== null) {
          // New page rendering is pending
          this._drawPage(this._pageNumPending);
          this._pageNumPending = null;
        }
      })
  },

  _handleTrack(evt) {
    let tmp;
    let getDiff = (evt) => {
      return {
        x: this._trackPos.x - evt.detail.x,
        y: this._trackPos.y -  evt.detail.y,
      };
    };
    switch(evt.detail.state) {
      case 'start':
       this._trackPos = { x:evt.detail.x, y:evt.detail.y};
       break;
     case 'track':
       tmp = getDiff(evt);
       this.$.viewer.style.transform = `translate(${this._pos.x-tmp.x}px,${this._pos.y-tmp.y}px )`;
       break;
     case 'end':
       tmp = getDiff(evt);
       this._pos.x = this._pos.x-tmp.x;
       this._pos.y = this._pos.y-tmp.y;
       this._recenter();
       break;
   }
 },

  _recenter() {
    const rect = this.$.container.getBoundingClientRect();
    const viewerRect = this.$.viewer.getBoundingClientRect();
    if(rect.width > viewerRect.width) {
      this._pos.x = (rect.width - viewerRect.width) / 2;
    }
    if(rect.height > viewerRect.height) {
      this._pos.y = (rect.height - viewerRect.height) / 2;
    }
    if( rect.width < viewerRect.width && rect.width > viewerRect.width + this._pos.x + 50) {
      this._pos.x = rect.width - 50 - viewerRect.width;
    }
    if( rect.width < viewerRect.width && this._pos.x > 50 ) {
      this._pos.x = 50
    }
    if( rect.height < viewerRect.height && rect.height > viewerRect.height + this._pos.y + 50) {
      this._pos.y = rect.height - 50 - viewerRect.height;
    }
    if( rect.height < viewerRect.height && this._pos.y > 50 ) {
      this._pos.y = 50
    }
    this.$.viewer.style.transform = `translate(${this._pos.x}px,${this._pos.y}px )`;
  }

});


components/pdf-viewer/pdf-viewer.es5.js
'use strict';

Polymer({

  is: 'pdf-viewer',

  properties: {
    /**
     * url of the pdf file to display
     * @type {String}
     */
    src: {
      type: String,
      observer: '_srcChanged'
    },

    /**
     * The pqge number to display
     *
     * The first page is 1
     * @type {Number}
     */
    page: {
      type: Number,
      value: 1,
      observer: '_pageChanged',
      reflectToAttribute: true,
      notify: true
    },

    /**
     * total number of pqges in the document
     * @type {Number}
     */
    pages: {
      type: Number,
      readOnly: true,
      reflectToAttribute: true,
      notify: true
    },

    /**
     * the initial zoom when opening a document and
     * when the mode change
     *
     * the available values are :
     *  - `fit` : whole page,
     *  - `fit-width` : full width
     *
     * @type {String}
     */
    initialZoom: {
      type: String,
      value: "fit"
    },

    /**
     * The zoom ratio used
     * @type {Numder}
     */
    zoomRatio: {
      type: Number,
      value: 1.25
    },

    /**
     * The view mode of the document
     *
     * the available values are :
     *  - single : single page
     *  - double : double page side by side
     *
     * @type {String}
     */
    mode: {
      type: String,
      value: "single",
      observer: '_modeChanged'
    },

    _views: {
      type: Number,
      value: 1
    },
    _PDF: {
      type: Object,
      value: function value() {
        return {};
      }
    },
    _pageRendering: {
      type: Boolean,
      value: false
    },
    _pageNumPending: {
      type: Number,
      value: null
    },
    _firstRender: {
      type: Boolean,
      value: false
    },
    _pos: {
      type: Object,
      value: function value() {
        return { x: 0, y: 0 };
      }
    },
    _zoom: {
      type: Number,
      value: 1
    }
  },

  attached: function attached() {
    var _this = this;

    this._ro = new ResizeObserver(function (entries, observer) {
      console.log('resize detect');
      if (_this._firstRender) _this._drawPage();
    });
    this._ro.observe(this);
  },
  detached: function detached() {
    delete this._ro;
  },
  _srcChanged: function _srcChanged() {
    var _this2 = this;

    this.firstRender = false;
    this.page = 1;
    PDFJS.getDocument(this.src).then(function (pdf) {
      _this2._setPages(pdf.numPages);
      _this2._PDF = pdf;
      switch (_this2.initialZoom) {
        case "fit-width":
          _this2.fitWidth();
          break;
        default:
          _this2.fit();
      }
    });
  },
  _pageChanged: function _pageChanged(newValue, oldValue) {
    if (!this.pages) {
      this.page = 1;
    }
    if (newValue > this.pages) this.page = this.pages;
    if (newValue < 1) this.page = 1;
    if (this._PDF) this._drawPage();
  },
  _modeChanged: function _modeChanged(newValue, oldValue) {
    switch (this.mode) {
      case 'double':
        this._views = 2;
        this.$.viewer2.hidden = false;
        break;
      default:
        this._views = 1;
        this.$.viewer2.hidden = true;
    }
    if (this._firstRender) {
      this._recenter();
      switch (this.initialZoom) {
        case "fit-width":
          this.fitWidth();
          break;
        default:
          this.fit();
      }
    }
  },


  /**
   * move to the next page
   */
  next: function next() {
    if (this.page < this.pages) this.page += this._views;
  },


  /**
   * move to the previous page
   */
  previous: function previous() {
    if (this.page > 1) this.page -= this._views;
  },


  /**
   * display the document full width
   */
  fitWidth: function fitWidth() {
    var _this3 = this;

    this._PDF.getPage(this.page).then(function (page) {
      var viewport = page.getViewport(1);
      var rect = _this3.$.container.getBoundingClientRect();
      _this3._zoom = (rect.width - 20) / (_this3._views * viewport.width);
      if (!_this3._firstRender) {
        viewport = page.getViewport(_this3._zoom);
        _this3._pos.x = (rect.width - _this3._views * viewport.width) / 2;
        _this3._pos.y = 10;
        _this3.$.viewer.style.transform = 'translate(' + _this3._pos.x + 'px,' + _this3._pos.y + 'px )';
      }
      _this3._drawPage();
    });
  },


  /**
   * display the whole page
   */
  fit: function fit() {
    var _this4 = this;

    this._PDF.getPage(this.page).then(function (page) {
      var viewport = page.getViewport(1);
      var rect = _this4.$.container.getBoundingClientRect();
      _this4._zoom = Math.min((rect.width - 20) / _this4._views * viewport.width, (rect.height - 20) / viewport.height);
      if (!_this4._firstRender) {
        viewport = page.getViewport(_this4._zoom);
        _this4._pos.x = (rect.width - _this4._views * viewport.width) / 2;
        _this4._pos.y = (rect.height - viewport.height) / 2;
        _this4.$.viewer.style.transform = 'translate(' + _this4._pos.x + 'px,' + _this4._pos.y + 'px )';
      }
      _this4._drawPage();
    });
  },


  /**
   * zoom in
   */
  zoomin: function zoomin() {
    this._zoom = this._zoom * this.zoomRatio;
    this._drawPage();
  },


  /**
   * zoom out
   */
  zoomout: function zoomout() {
    this._zoom = this._zoom / this.zoomRatio;
    this._drawPage();
  },
  _renderView: function _renderView(view, pg) {
    var _this5 = this;

    return new Promise(function (resolve, reject) {
      _this5._PDF.getPage(pg).then(function (page) {
        var viewport = page.getViewport(_this5._zoom);
        view.width = viewport.width;
        view.height = viewport.height;

        var context = view.getContext('2d');
        var renderContext = {
          canvasContext: context,
          viewport: viewport
        };
        resolve(page.render(renderContext));
      }).catch(function (err) {
        _this5._PDF.getPage(1).then(function (page) {
          var viewport = page.getViewport(_this5._zoom);
          view.width = viewport.width;
          view.height = viewport.height;
        }).then(resolve);
      });
    });
  },
  _drawPage: function _drawPage(pg) {
    var _this6 = this;

    if (!pg) pg = this.page;
    if (this._pageRendering) {
      this._pageNumPending = pg;
      return;
    }
    this._pageRendering = true;
    this.$.spinner.active = true;
    var promises = [];
    if (this.mode === "single") {
      promises.push(this._renderView(this.$.viewer1, pg));
    } else {

      promises.push(this._renderView(this.$.viewer1, pg - pg % 2));
      promises.push(this._renderView(this.$.viewer2, pg - pg % 2 + 1));
    }
    Promise.all(promises).then(function () {
      _this6._firstRender = true;
      _this6._pageRendering = false;
      _this6.$.spinner.active = false;
      _this6._recenter();
      if (_this6._pageNumPending !== null) {
        // New page rendering is pending
        _this6._drawPage(_this6._pageNumPending);
        _this6._pageNumPending = null;
      }
    });
  },
  _handleTrack: function _handleTrack(evt) {
    var _this7 = this;

    var tmp = void 0;
    var getDiff = function getDiff(evt) {
      return {
        x: _this7._trackPos.x - evt.detail.x,
        y: _this7._trackPos.y - evt.detail.y
      };
    };
    switch (evt.detail.state) {
      case 'start':
        this._trackPos = { x: evt.detail.x, y: evt.detail.y };
        break;
      case 'track':
        tmp = getDiff(evt);
        this.$.viewer.style.transform = 'translate(' + (this._pos.x - tmp.x) + 'px,' + (this._pos.y - tmp.y) + 'px )';
        break;
      case 'end':
        tmp = getDiff(evt);
        this._pos.x = this._pos.x - tmp.x;
        this._pos.y = this._pos.y - tmp.y;
        this._recenter();
        break;
    }
  },
  _recenter: function _recenter() {
    var rect = this.$.container.getBoundingClientRect();
    var viewerRect = this.$.viewer.getBoundingClientRect();
    if (rect.width > viewerRect.width) {
      this._pos.x = (rect.width - viewerRect.width) / 2;
    }
    if (rect.height > viewerRect.height) {
      this._pos.y = (rect.height - viewerRect.height) / 2;
    }
    if (rect.width < viewerRect.width && rect.width > viewerRect.width + this._pos.x + 50) {
      this._pos.x = rect.width - 50 - viewerRect.width;
    }
    if (rect.width < viewerRect.width && this._pos.x > 50) {
      this._pos.x = 50;
    }
    if (rect.height < viewerRect.height && rect.height > viewerRect.height + this._pos.y + 50) {
      this._pos.y = rect.height - 50 - viewerRect.height;
    }
    if (rect.height < viewerRect.height && this._pos.y > 50) {
      this._pos.y = 50;
    }
    this.$.viewer.style.transform = 'translate(' + this._pos.x + 'px,' + this._pos.y + 'px )';
  }
});
